### PR TITLE
[NHWC SUPPORT] Fix suggest_memory_format() for ambiguous cases

### PR DESCRIPTION
**Pull request by jjsjann123**
This PR is to fix the ambiguity in NC11 and N1HW tensor when specifying
memory_format.
Previous implementation unnecessarily associate memory contiguity with memory
format, which undermines the expressiveness of memory_format for the ambiguous
cases.

e.g. Previously, if we have NC11 tensor in channels_last format. Calling
contiguous on that tensor doesn't change the memory_format to NCHW.

### CODE REVIEW

**Comment by jjsjann123** [comment]
cc'ing @csarofeen @ptrblck for visibility

File: aten/src/ATen/test/mem_overlap_test.cpp [hunk 1]
```diff
@@ -0,0 +1,61 @@
+#include <gtest/gtest.h>
```
Comment (VitalyFedyunin) [inline aten/src/ATen/test/mem_overlap_test.cpp]: Why it is part of PR?

Comment (jjsjann123) [reply aten/src/ATen/test/mem_overlap_test.cpp]: My bad, switching branch doesn't get rid of files that's not tracked in this one.... Will remove this file

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by jjsjann123** [comment]
Failure looks related. Looking into that now.

**Comment by jjsjann123** [comment]
The issue has been fixed. However, the native conv transpose back looks suspicious to me.

In `derivatives.yaml`, we are preserving memory_format of grad on the column: https://github.com/pytorch/pytorch/blob/master/tools/autograd/derivatives.yaml#L1182

But the backward function actually requires column to be contiguous:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/NaiveConvolutionTranspose2d.cpp#L790
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/NaiveConvolutionTranspose2d.cpp#L411-L412

Looks like we have a bug in the code.
I do not know whether the kernel handles channels last at all, so I don't know the right fix here (whether to update the check / or remove the memory_format preserve part).

File: aten/src/ATen/native/TensorProperties.cpp [hunk 2]
```diff
@@ -77,7 +77,14 @@ Tensor contiguous(const Tensor & self) {
...
 Tensor contiguous(const Tensor& self, MemoryFormat memory_format) {
-  if (self.is_contiguous(memory_format)) {
+  // Possibly that contiguous tensor is not strided in memory_format.
+  // so later `suggest_memory_layout` does not reflect the memory layout update.
+  // The alternative to making a copy here is to duplicate `TensorImpl`
+  // inside self and restride it, wrap that with another `Tensor` and return it;
+  // I'm concerned that now we have two TensorImpl sharing the same storage,
+  // which would potentially invalidate each other's `version_counter_`.
+  // Let's just make a copy here for now.
+  if (self.is_contiguous(memory_format) && self.suggest_memory_format() == memory_format) {
```
Comment (jjsjann123) [inline aten/src/ATen/native/TensorProperties.cpp]: As @ngimel kindly pointed out, this copy could be avoided.
`as_restride` does create new TensorImpl that shared version_counter, hence my concern is not legit. I'll update the code

**Comment by ngimel** [comment]
> 
> Looks like we have a bug in the code.
> I do not know whether the kernel handles channels last at all, so I don't know the right fix here (whether to update the check / or remove the memory_format preserve part).
Native convolution implementations don't handle non-contiguous input, so memory_format preserve part should be removed.

File: test/test_nn.py [hunk 3]
```diff
@@ -3529,6 +3529,27 @@ def test_adaptive_pooling_avg_nhwc(self):
         self.assertEqual(out, ref_out)
         self.assertEqual(input.grad, ref_input.grad)
 
+    @unittest.skipIf(not TEST_CUDA, "CUDA unavailable")
```
Comment (ngimel) [inline test/test_nn.py]: why are you testing only cuda here? There's no implementation on cpu?
Can you please move your tests to the generic test case (class TestNNDeviceType, on line 8575 approx) and use generic framework for them https://github.com/pytorch/pytorch/wiki/Writing-tests-that-run-on-all-available-device-types

File: test/test_nn.py [hunk 4]
```diff
@@ -3529,6 +3529,27 @@ def test_adaptive_pooling_avg_nhwc(self):
...
+        input = input.contiguous(memory_format=torch.channels_last).requires_grad_()
+        grad = torch.randint(1, 10, (4, 8, 4, 4), dtype=torch.float32, device="cuda")
+        pool = torch.nn.AdaptiveAvgPool2d((4, 4))
+
+        ref_input = input.detach().clone().contiguous().requires_grad_(True)
+        ref_grad = grad.detach().clone().contiguous()
+        ref_pool = torch.nn.AdaptiveAvgPool2d((4, 4)).cuda()
+
+        out = pool(input)
+        out.backward(grad)
```
Comment (ngimel) [inline test/test_nn.py]: How does backward work with contiguous (not channels-last) grad?

File: test/test_nn.py [hunk 5]
```diff
@@ -7471,6 +7492,36 @@ def test_conv_cudnn_nhwc(self):
         self.assertEqual(conv.bias.grad, ref_conv.bias.grad)
         self.assertEqual(input.grad, ref_input.grad)
 
+    @unittest.skipIf(not TEST_CUDA, "CUDA unavailable")
+    @unittest.skipIf(not TEST_CUDNN, "needs cudnn")
+    @skipIfRocm
+    def test_conv_cudnn_nhwc_size_1_filter(self):
```
Comment (ngimel) [inline test/test_nn.py]: Same here, please move into device-generic class, and use new-style decorators to run only on cuda with cudnn

File: test/test_nn.py [hunk 6]
```diff
@@ -7471,6 +7492,36 @@ def test_conv_cudnn_nhwc(self):
...
+        out.backward(grad)
+        ref_out = ref_conv(ref_input)
+        ref_out.backward(ref_grad)
+
+        self.assertTrue(out.is_contiguous(memory_format=torch.channels_last))
+        self.assertTrue(ref_out.is_contiguous())
+        self.assertEqual(out, ref_out)
+        self.assertEqual(conv.weight.grad, ref_conv.weight.grad)
+        self.assertEqual(conv.bias.grad, ref_conv.bias.grad)
+        self.assertEqual(input.grad, ref_input.grad)
```
Comment (ngimel) [inline test/test_nn.py]: why are you not asserting input.grad memory format?

**Comment by jjsjann123** [comment]
@ngimel Thanks for confirming that. I'll do a separate PR on that, as it's not related here.

**Review by ngimel — CHANGES_REQUESTED** [review]

File: test/test_nn.py (hunk 6 above)
Comment (jjsjann123) [reply test/test_nn.py]: Because of this guy: https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/functions/accumulate_grad.cpp#L60

As we slowly merge things back from the branch, it will eventually be fixed here. But I think that part is not merged to master yet.

File: test/test_nn.py (hunk 4 above)
Comment (jjsjann123) [reply test/test_nn.py]: I will add a test here.

Kernel would output grad_input as the same memory_format with input, handles grad_output as it (so memory accessing pattern will be bad, but my intuition is, it should still be better than another permutation kernel, I should have benchmarked it :smile: )

File: test/test_nn.py (hunk 3 above)
Comment (jjsjann123) [reply test/test_nn.py]: No there's no CPU NHWC support yet.
Will move and update decorator as instructed.

File: c10/core/TensorImpl.cpp [hunk 7]
```diff
@@ -116,14 +116,20 @@ bool TensorImpl::compute_channels_last_contiguous() const {
...
+      return false;
+    }
     for (auto& d : {1, 3, 2, 0}) {
-      if (size(d) != 1) {
-        if (stride(d) > min) {
-          min = stride(d);
-        } else {
-          return false;
-        }
+      if (stride(d) < min) {
```
[the remark this answers is not in the document]
Comment (ngimel) [reply c10/core/TensorImpl.cpp]: Ok, Jie convinced me offline that he needs 1-dim strides to carry payload. Much as I hate it and as fragile as it is, it can stay here for now.

**Comment by VitalyFedyunin** [comment]
> The issue has been fixed. However, the native conv transpose back looks suspicious to me.
> 
> In `derivatives.yaml`, we are preserving memory_format of grad on the column: [/tools/autograd/derivatives.yaml@`master`#L1182](https://github.com/pytorch/pytorch/blob/master/tools/autograd/derivatives.yaml#L1182) [](/pytorch/pytorch/blob/HEAD@{2019-11-21T00:19:42Z}/tools/autograd/derivatives.yaml#L1182)
> 
> But the backward function actually requires column to be contiguous:
> [/aten/src/ATen/native/NaiveConvolutionTranspose2d.cpp@`master`#L790](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/NaiveConvolutionTranspose2d.cpp#L790) [](/pytorch/pytorch/blob/HEAD@{2019-11-21T00:19:42Z}/aten/src/ATen/native/NaiveConvolutionTranspose2d.cpp#L790)
> [/aten/src/ATen/native/NaiveConvolutionTranspose2d.cpp@`master`#L411-L412](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/NaiveConvolutionTranspose2d.cpp#L411-L412) [](/pytorch/pytorch/blob/HEAD@{2019-11-21T00:19:42Z}/aten/src/ATen/native/NaiveConvolutionTranspose2d.cpp#L411-L412)
> 
> Looks like we have a bug in the code.
> I do not know whether the kernel handles channels last at all, so I don't know the right fix here (whether to update the check / or remove the memory_format preserve part).

Indeed CPU backward (as well as forward) will turn tensor into contiguous format. I don't see any bug here, just implementation opportunity, correct me if I'm worng.

**Comment by jjsjann123** [comment]
> Indeed CPU backward (as well as forward) will turn tensor into contiguous format. I don't see any bug here, just implementation opportunity, correct me if I'm worng.

I think it's a bug here:
`derivatives.yaml` construct `grad_column` under `empty_like(grad, at::MemoryFormat::Preserve)`, where `grad` was passed from the following layer. (https://github.com/pytorch/pytorch/blob/master/tools/autograd/derivatives.yaml#L1182)

So if the next layer passed a gradient function as NHWC, `slow_conv_transpose2d_backward` would have `grad_column` in NHWC, which the kernel can't handle and errors out at https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/NaiveConvolutionTranspose2d.cpp#L411-L412

**Comment by jjsjann123** [comment]
I have a coming PR that fixed the cudnn descriptor, which unfortunately has dependency on this one. Should I wait for this to be merged first or stash it on top so we have it all fixed in one go?
